lib/simple_abs.rb

```diff
@@ -2,110 +2,91 @@
+    return params[ab_test_name] if params[ab_test_name].present?
 
-    test_value = cookies[name]
+    # Now check of cookie exists
+    test_json = ab_test_read_cookie(ab_test_name)
 
-    if test_value.blank? || !tests.include?(test_value)
+    if !test_json
       test_value = tests[rand(tests.size)]
-      cookies.permanent[name] = test_value
+      test_record = AbTest.create!(experiment: name, choice: test_value, impression: impression)
+      ab_test_save_cookie(ab_test_name, test_record)
+    elsif !test_json[:finished]
+      test_record = AbTest.find(test_json[:id])
+      test_value = test_record.choice rescue test_json[:choice]
+      test_record.increment!(:impression, impression)
+      ab_test_save_cookie(ab_test_name, test_record)
+    else # Converted
+      test_record = AbTest.find(test_json[:id])
+      test_value = test_record.choice rescue test_json[:choice]
     end
 
-    find_or_create_by_experiment_and_which_method(name, test_value).increment!(:participants, impression)
-
     return test_value
   end
 
-  def converted!(name)
-
-    if !is_bot?
-      test_value = cookies[name]
-      if test_value && cookies[name.to_s + "_converted"].blank?
-        find_or_create_by_experiment_and_which_method(name, test_value).increment!(:conversions)
-        cookies.permanent[name.to_s + "_converted"] = true
-      end
-    end
-  end
-
-  def converted_with_frequency!(name, frequency = 1)
-
-    if !is_bot?
-      test_value = cookies[name]
-      if test_value
-        find_or_create_by_experiment_and_which_method(name, test_value).increment!(:conversions, frequency)
+  # For converted, we have three scenarios
+  # 0. Not a participant, ignore
+  # 1. Already converted, do nothing
+  # 2. Conversion with no finsih: Add 1/N to conversion, and tag time if first
+  # 3. Conversion with finish: Do #2 and set cookie to be finished.
+  def ab_test_converted!(name, conversion: 1, finished: true)
+    if !browser.bot?
+      ab_test_name = ('ab_test_' + name).to_sym
+      test_json = ab_test_read_cookie(ab_test_name)
+      if test_json && !test_json[:finished]
+        test_record = AbTest.find(test_json[:id])
+        test_record.increment!(:conversion, conversion)
+        test_record.update_attribute(:converted_at, Time.now()) unless test_record.converted_at.present?
+        if finished
+          ab_test_save_cookie(ab_test_name, test_record, 'converted')
+        else
+          ab_test_save_cookie(ab_test_name, test_record)
+        end
       end
     end
   end
 
   def ab_test_aborted!(name)
-
-    if !is_bot?
-      test_value = cookies[name]
-      if test_value && cookies[name.to_s + "_converted"].blank?
-        find_or_create_by_experiment_and_which_method(name, test_value).decrement!(:participants)
-        cookies.permanent[name.to_s + "_converted"] = true
+    if !browser.bot?
+      ab_test_name = ('ab_test_' + name).to_sym
+      test_json = ab_test_read_cookie(ab_test_name)
+      test_value = test_json
```

Comment:
Earlier we have used test_value = test_json[:choice], so it seems strange to set these equal here.
I'd rather just remove this reference to test_value, and write things here in terms of test_json.